contracts/MsgOracle.sol

```diff
@@ -0,0 +1,83 @@
+pragma solidity ^0.5.12;
+
+import "node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+/**
+@title MsgOracle
+@author Rinke Hendriksen <rinke@ethswarm.org>
+@notice Set the honey prices for various messages in Swarm
+@dev This on-chain msgOracle must be queried by a swarm-node. By reading the event logs, a node is able to construct the following object:
+msgPrices: {TTL, prices: {swarmMsgX: [<validFrom0, price>, <validFromN, price>], swarmMsgX: [<validFrom0, price>, <validFromN, price>]}}.
+this object expires after TTL seconds. The prices object holds all message types with their respective price arrays.
+In the price array of a message, multiple prices can be quoted. The chosen price is the most-recent price.
+*/
+contract MsgOracle is Ownable {
+    /*
+    TTL is the TTL applied by nodes (how long is a query to this oracle valid)
+    oldTTL is used internally when lastUpdated is not longer than oldTTL ago,
+    furthermore, we don't allow updating TTL when lastUpdated is not TTL seconds ago.
+    */
+    uint256 public TTL;
+    uint256 private oldTTL;
+    uint256 private lastUpdated;
+
+    event LogNewTTL(uint256 TTL);
+    event LogSetMsgPrice(bytes32 indexed swarmMsg, uint256 price, uint256 validFrom);
+    event LogRevertMsgPrice(bytes32 indexed swarmMsg, uint256 price, uint256 validFrom);
+    /**
+    @notice Sets an initial value of TTL and sets the owner
+    @dev The owner can be a EOA or a smart-contract (with any arbitrary governance structure) which can call all functions from this contract
+    @param _TTL The initial TTL, effective immediately.
+    */
+    constructor(uint256 _TTL) public Ownable() {
+        TTL = _TTL;
+        lastUpdated = now;
+        emit LogNewTTL(TTL);
+    }
+
+    /**
+    @notice Sets a new value of TTL.
+    @dev The TTL is the duration in seconds of the validity of a query to this oracle.
+    We advise querying the oracle again at a fraction of TTL (e.g. 50%).
+    Sets the TTL from the msgPrices object {TTL, prices}
+    Warning: the owner is responsible for ensuring that we don't update TTL to an unreasonably high value and hereby deadlock the contract for this period.
+    @param _TTL TTL which will be effective after TTL seconds
+    */
+    function newTTL(uint256 _TTL) public onlyOwner {
```

Comment:
What prevents the following scenario:

* At time `t` TTL is 600s and was set 600 seconds ago.
* Node A queries the oracle at `t+1` seeing a TTL of 600 seconds
* At time `t+2` `newTTL` is called with 1s. (`oldTTL=600` and `TTL=1`)
* At time `t+3` `newTTL` is called with 1s. (`oldTTL=1` and `TTL=1`)
* At time `t+4` `setMsgPrice` is called with `validFrom` being `t+5` (this would work because `validFrom >= now + TTL` would be `t+5 >= t+4+1`)
* Nodes which query the price now will see the new price starting from `t+5` but node A still has the old price and won't requery until `t+301` (if the 50% recommendation is followed).

Reply:
Nothing. I propose to change the body of newTTL to:
```solidity
    if(lastUpdated > now - oldTTL) {
            require(lastUpdated.add(oldTTL) <= now, "MsgOracleOwner: TTL less than oldTTL seconds ago updated");
        } else {
            require(lastUpdated.add(TTL) <= now, "MsgOracleOwner: TTL less than TTL seconds ago updated");
        }
        oldTTL = TTL;
        TTL = _TTL;
        lastUpdated = now;
        emit LogNewTTL(_TTL);
```

Reply:
The logic of that seems strange. `lastUpdated > now - oldTTL` is equivalent to `lastUpdated + oldTTL > now` and if thats true we check that `lastUpdated + oldTTL <= now`.

Reply:
Nothing yet. I propose to update it to: 
```solidity
        if(lastUpdated > now - oldTTL) {
            revert("MsgOracleOwner: oldTTL still active");
        } else {
            require(lastUpdated.add(TTL) <= now, "MsgOracleOwner: TTL less than TTL seconds ago updated");
        }
        oldTTL = TTL;
        TTL = _TTL;
        lastUpdated = now;
        emit LogNewTTL(_TTL);
```

Reply:
That might work. I would rewrite it that way (should be equivalent to your code):
```solidity
require(lastUpdated.add(oldTTL) <= now, "oldTTL still active");
require(lastUpdated.add(TTL) <= now, "MsgOracleOwner: TTL less than TTL seconds ago updated");
oldTTL = TTL;
TTL = _TTL;
lastUpdated = now;
emit LogNewTTL(_TTL);
```

Reply:
> The logic of that seems strange. lastUpdated > now - oldTTL is equivalent to lastUpdated + oldTTL > now and if thats true we check that lastUpdated + oldTTL <= now.

Yes, sorry. This was strange indeed. I deleted that comment, as it was wrong.